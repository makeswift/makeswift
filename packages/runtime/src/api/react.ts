import { type FetchableValue } from '@makeswift/controls'
import * as MakeswiftApiClient from '../state/makeswift-api-client'
import {
  APIResourceType,
  File,
  GlobalElement,
  LocalizedGlobalElement,
  Page,
  PagePathnameSlice,
  Site,
  Snippet,
  Swatch,
  Table,
  Typography,
} from './types'
import { GraphQLClient } from './graphql/client'
import { CreateTableRecordMutation } from './graphql/documents'
import {
  CreateTableRecordMutationResult,
  CreateTableRecordMutationVariables,
} from './graphql/generated/types'

export type SerializedLocalizedResourcesMap = {
  [resourceId: string]: string | null
}

type LocalizedResourcesMap = Map<string, string | null>

export type CacheData = MakeswiftApiClient.SerializedState

export type MakeswiftClientOptions = {
  uri: string
  cacheData?: CacheData
  localizedResourcesMap?: SerializedLocalizedResourcesMap
  locale?: Intl.Locale
}

/**
 * NOTE(miguel): This "client" is used to fetch Makeswift API resources needed for the host. For
 * example, swatches, files, typographies, etc. Ideally it's internal to the runtime and is only
 * used by controls to transform API references to API resources.
 *
 * Moreover, its use should be reserved for the builder only, since for live pages all Makeswift
 * API resources should be embedded in the "page snapshot". In the builder, this client serves the
 * purpose of sending requests for API resources and keeping a cache so that changes that happen in
 * the builder, like modifying a swatch, can be sent via `postMessage` to the host and the cache can
 * immediately update the value and re-render.
 *
 * Furthermore, the API resources requested shouldn't be requested directly from the Makeswift API
 * as that would require those resources to not be authenticated since the requests come from the
 * browser when running the host. Instead, the requests should go to the host directly, at the
 * Makeswift API endpoint (i.g., `/api/makeswift/[...makeswift]` dynamic route) where the host's
 * API key can be used, securely, in the server. For this reason, this client should really be a
 * client of the host's API, not Makeswift's, intended to build and continuously maintain a realtime
 * snapshot for use in the builder, not the lives pages.
 */
export class MakeswiftHostApiClient {
  graphqlClient: GraphQLClient
  makeswiftApiClient: MakeswiftApiClient.Store
  subscribe: MakeswiftApiClient.Store['subscribe']
  private localizedResourcesMap: LocalizedResourcesMap
  private locale: Intl.Locale | undefined

  constructor({ uri, cacheData, localizedResourcesMap = {}, locale }: MakeswiftClientOptions) {
    this.graphqlClient = new GraphQLClient(uri)
    this.makeswiftApiClient = MakeswiftApiClient.configureStore({ serializedState: cacheData })
    this.subscribe = this.makeswiftApiClient.subscribe
    this.localizedResourcesMap = new Map(Object.entries(localizedResourcesMap))
    this.locale = locale
  }

  readSwatch(swatchId: string): Swatch | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Swatch,
      swatchId,
    )
  }

  async fetchSwatch(swatchId: string): Promise<Swatch | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Swatch, swatchId),
    )
  }

  resolveSwatch(swatchId: string | undefined): FetchableValue<Swatch | null> {
    return this.resolveResource(APIResourceType.Swatch, {
      id: swatchId,
      read: id => this.readSwatch(id),
      fetch: id => this.fetchSwatch(id),
    })
  }

  readFile(fileId: string): File | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.File,
      fileId,
    )
  }

  async fetchFile(fileId: string): Promise<File | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.File, fileId),
    )
  }

  resolveFile(fileId: string | undefined): FetchableValue<File | null> {
    return this.resolveResource(APIResourceType.File, {
      id: fileId,
      read: id => this.readFile(id),
      fetch: id => this.fetchFile(id),
    })
  }

  readTypography(typographyId: string): Typography | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Typography,
      typographyId,
    )
  }

  async fetchTypography(typographyId: string): Promise<Typography | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Typography, typographyId),
    )
  }

  resolveTypography(typographyId: string | undefined): FetchableValue<Typography | null> {
    return this.resolveResource(APIResourceType.Typography, {
      id: typographyId,
      read: id => this.readTypography(id),
      fetch: id => this.fetchTypography(id),
    })
  }

  readGlobalElement(globalElementId: string): GlobalElement | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.GlobalElement,
      globalElementId,
    )
  }

  async fetchGlobalElement(globalElementId: string): Promise<GlobalElement | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.GlobalElement, globalElementId),
    )
  }

  readLocalizedGlobalElement(globalElementId: string): LocalizedGlobalElement | null {
    const localizedGlobalElementId = this.getLocalizedResourceId(globalElementId)

    if (localizedGlobalElementId == null) return null

    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.LocalizedGlobalElement,
      localizedGlobalElementId,
    )
  }

  async fetchLocalizedGlobalElement(
    globalElementId: string,
  ): Promise<LocalizedGlobalElement | null> {
    const locale = this.locale

    if (locale == null) return null

    // If getLocalizedResourceId returns null, it means we have tried to fetch the resource,
    // but the resource is not available. If we haven't fetched it yet, it'll return undefined.
    const noLocalizedResource = this.getLocalizedResourceId(globalElementId) === null

    if (noLocalizedResource) return null

    // TODO(fikri): We're checking the cache here because the cache hit will fail on fetchAPIResource.
    // This is because in the cache we're saving the ID of the localizedGlobalElementId,
    // but we're reading the cache using the globalElementId.
    // This is very confusing and it can lead to more bugs. We need to refactor how we handle
    // localized global element.
    const cacheResult = this.readLocalizedGlobalElement(globalElementId)

    if (cacheResult) return cacheResult

    const result = await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(
        APIResourceType.LocalizedGlobalElement,
        globalElementId,
        locale,
      ),
    )

    this.setLocalizedResourceId({
      resourceId: globalElementId,
      localizedResourceId: result?.id ?? null,
    })

    return result
  }

  readPagePathnameSlice(pageId: string): PagePathnameSlice | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.PagePathnameSlice,
      pageId,
    )
  }

  async fetchPagePathnameSlice(pageId: string): Promise<PagePathnameSlice | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.PagePathnameSlice, pageId, this.locale),
    )
  }

  resolvePagePathnameSlice(pageId: string | undefined): FetchableValue<PagePathnameSlice | null> {
    return this.resolveResource(APIResourceType.PagePathnameSlice, {
      id: pageId,
      read: id => this.readPagePathnameSlice(id),
      fetch: id => this.fetchPagePathnameSlice(id),
    })
  }

  resolveResource<R>(
    type: APIResourceType,
    {
      id,
      read,
      fetch,
    }: {
      id: string | undefined
      read: (id: string) => R | null
      fetch: (id: string) => Promise<R | null>
    },
  ): FetchableValue<R | null> {
    const _read = () => (id != null ? read(id) : null)
    let lastValue: R | null = null
    return {
      name: `${type}:${id}`,
      readStable: () => (lastValue = _read()),
      subscribe: (onUpdate: () => void) =>
        this.subscribe(() => {
          if (_read() !== lastValue) onUpdate()
        }),
      fetch: async () => (id != null ? fetch(id) : null),
    }
  }

  readTable(tableId: string): Table | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Table,
      tableId,
    )
  }

  async fetchTable(tableId: string): Promise<Table | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Table, tableId),
    )
  }

  async createTableRecord(tableId: string, columns: any): Promise<void> {
    await this.graphqlClient.request<
      CreateTableRecordMutationResult,
      CreateTableRecordMutationVariables
    >(CreateTableRecordMutation, { input: { data: { tableId, columns } } })
  }

  readSite(siteId: string): Site | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Site,
      siteId,
    )
  }

  readPage(pageId: string): Page | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Page,
      pageId,
    )
  }

  readSnippet(snippetId: string): Snippet | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Snippet,
      snippetId,
    )
  }

  private getLocalizedResourceId(resourceId: string): string | undefined | null {
    return this.localizedResourcesMap?.get(resourceId)
  }

  setLocalizedResourceId({
    resourceId,
    localizedResourceId,
  }: {
    resourceId: string
    localizedResourceId: string | null
  }): void {
    this.localizedResourcesMap.set(resourceId, localizedResourceId)
  }
}
