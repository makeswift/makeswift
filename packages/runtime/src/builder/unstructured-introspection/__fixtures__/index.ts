export * from './element-containing-resources'
