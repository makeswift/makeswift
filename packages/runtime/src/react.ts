export type { Action } from './state/actions'
export { ActionTypes, changeDocument } from './state/actions'
export {
  Element,
  ReactRuntime,
  useIsInBuilder,
  DocumentRoot,
  PageProvider,
  usePageId,
} from './runtimes/react'
