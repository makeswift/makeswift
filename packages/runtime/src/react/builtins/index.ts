export { MakeswiftComponentType } from '../../components/builtin/constants'
