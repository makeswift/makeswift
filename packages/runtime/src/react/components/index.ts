/**
 * Re-export all component-related code
 */
export * from './base-image';
export * from './base-link';