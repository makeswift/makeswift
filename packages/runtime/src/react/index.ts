export { ReactRuntime } from '../runtimes/react'
