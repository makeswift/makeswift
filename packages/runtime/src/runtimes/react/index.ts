export { ReactRuntime } from './react-runtime'

export { Element } from './components/Element'
export { DocumentRoot } from './components/DocumentRoot'
export { PageProvider } from './components/PageProvider'
export { ReactRuntimeProvider } from './components/RuntimeProvider'

export { useBreakpoints } from './hooks/use-breakpoints'
export { useBuilderEditMode } from './hooks/use-builder-edit-mode'
export { useDocumentKey } from './hooks/use-document-context'
export { useElementId } from './hooks/use-element-id'
export { useIsInBuilder } from './hooks/use-is-in-builder'
export { useIsPreview } from './hooks/use-is-preview'
export { usePageId } from './hooks/use-page-id'
export { useSelector } from './hooks/use-selector'
export { StoreContext, useStore } from './hooks/use-store'
