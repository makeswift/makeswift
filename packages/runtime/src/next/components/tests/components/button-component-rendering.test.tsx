/** @jest-environment jsdom */

import '@testing-library/jest-dom'
import { forwardRef } from 'react'

import {
  createMakeswiftPageSnapshot,
  testMakeswiftPageRendering,
  createRootComponent,
} from '../../../testing'

import { pageDocument } from '../__fixtures__/page-document'
import {
  buttonComponentData,
  linkData,
  linkUrlData,
} from '../__fixtures__/element-data/button-component'
import * as Resources from '../__fixtures__/resources'

jest.mock('next/link', () => ({
  __esModule: true,
  default: forwardRef<HTMLDivElement, any>(
    jest
      .fn()
      .mockImplementation(({ legacyBehavior, ...props }, ref) => (
        <a
          ref={ref}
          {...props}
          data-next-link="true"
          data-next-link-legacy-behavior={legacyBehavior}
        />
      )),
  ),
}))

const createPageWithButton = (...args: Parameters<typeof buttonComponentData>) => {
  const button = buttonComponentData(...args)

  return createMakeswiftPageSnapshot(
    {
      ...pageDocument,
      data: createRootComponent([button]),
    },
    {
      cacheData: {
        apiResources: {
          PagePathnameSlice: [
            {
              id: Resources.pagePathnameSlice1.id,
              value: Resources.pagePathnameSlice1,
            },
          ],
        },
      },
    },
  )
}

describe('correctly renders button component', () => {
  const htmlId = 'test-button'

  test('with link', async () => {
    const snapshot = createPageWithButton({
      htmlId,
      linkData: linkData.url,
    })

    const render = await testMakeswiftPageRendering({ snapshot })
    expect(render.container.querySelector(`#${htmlId}`)).toMatchSnapshot()
  })

  test.each([linkData.page, linkData.url, linkData.email, linkData.phone, linkData.element])(
    'with $type link',
    async linkData => {
      const snapshot = createPageWithButton({
        htmlId,
        linkData,
      })

      const render = await testMakeswiftPageRendering({ snapshot })
      const button = render.container.querySelector(`#${htmlId}`)
      button?.setAttribute('class', '[redacted]')
      expect(button).toMatchSnapshot()
    },
  )

  test.each([
    linkUrlData(''),
    linkUrlData('http:'),
    linkUrlData('http://'),
    linkUrlData('https:'),
    linkUrlData('https://'),
    linkUrlData('file:'),
    linkUrlData('file://'),
    linkUrlData('file://server/path'),
    linkUrlData('mailto:alan@makeswift.com'),
    linkUrlData('phone:+15551234567'),
  ])('with $type link $payload.url', async linkData => {
    const snapshot = createPageWithButton({
      htmlId,
      linkData,
    })

    const render = await testMakeswiftPageRendering({ snapshot })
    const button = render.container.querySelector(`#${htmlId}`)
    button?.setAttribute('class', '[redacted]')
    expect(button).toMatchSnapshot()
  })
})
