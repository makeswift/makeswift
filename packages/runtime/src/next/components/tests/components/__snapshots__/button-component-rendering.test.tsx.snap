// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders button component with CALL_PHONE link 1`] = `
<a
  class="[redacted]"
  href="tel:15551234567"
  id="test-button"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_PAGE link 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="/about"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link  1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href=""
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="https://www.makeswift.com/"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link file: 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="file:"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link file:// 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="file://"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link file://server/path 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="file://server/path"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link http: 1`] = `
<a
  class="[redacted]"
  href="http:"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link http:// 1`] = `
<a
  class="[redacted]"
  href="http://"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link https: 1`] = `
<a
  class="[redacted]"
  href="https:"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link https:// 1`] = `
<a
  class="[redacted]"
  href="https://"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link mailto:alan@makeswift.com 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="mailto:alan@makeswift.com"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with OPEN_URL link phone:+15551234567 1`] = `
<a
  class="[redacted]"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="phone:+15551234567"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;

exports[`correctly renders button component with SCROLL_TO_ELEMENT link 1`] = `
<a
  class="[redacted]"
  href="#test-button"
  id="test-button"
>
  Button Text
</a>
`;

exports[`correctly renders button component with SEND_EMAIL link 1`] = `
<a
  class="[redacted]"
  href="mailto:alan@makeswift.com?subject=Buttons&body="
  id="test-button"
>
  Button Text
</a>
`;

exports[`correctly renders button component with link 1`] = `
.emotion-0 {
  display: table;
  border: 0;
  outline: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
}

.emotion-1 {
  max-width: 100%;
}

@media only screen and (min-width: 769px) {
  .emotion-1 {
    width: auto;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-1 {
    width: auto;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-1 {
    width: auto;
  }
}

@media only screen and (min-width: 769px) {
  .emotion-2 {
    margin-top: 60px;
    margin-right: auto;
    margin-bottom: 80px;
    margin-left: auto;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-2 {
    margin-top: 60px;
    margin-right: auto;
    margin-bottom: 80px;
    margin-left: auto;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-2 {
    margin-top: 60px;
    margin-right: auto;
    margin-bottom: 80px;
    margin-left: auto;
  }
}

@media only screen and (min-width: 769px) {
  .emotion-3 {
    color: hsla(0,0%,100%,1);
    border-radius: 4px;
    padding: 12px 16px;
    font-size: 14px;
    background: hsla(0,0%,0%,1);
    border: none;
    -webkit-transition: color 0.15s ease-in-out,background 0.15s ease-in-out,border 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
    transition: color 0.15s ease-in-out,background 0.15s ease-in-out,border 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  }

  .emotion-3:hover {
    background: #000000;
  }

  .emotion-3:active {
    background: #000000;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-3 {
    color: hsla(0,0%,100%,1);
    border-radius: 4px;
    padding: 12px 16px;
    font-size: 14px;
    background: hsla(0,0%,0%,1);
    border: none;
    -webkit-transition: color 0.15s ease-in-out,background 0.15s ease-in-out,border 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
    transition: color 0.15s ease-in-out,background 0.15s ease-in-out,border 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  }

  .emotion-3:hover {
    background: #000000;
  }

  .emotion-3:active {
    background: #000000;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-3 {
    color: hsla(0,0%,100%,1);
    border-radius: 4px;
    padding: 12px 16px;
    font-size: 14px;
    background: hsla(0,0%,0%,1);
    border: none;
    -webkit-transition: color 0.15s ease-in-out,background 0.15s ease-in-out,border 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
    transition: color 0.15s ease-in-out,background 0.15s ease-in-out,border 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  }

  .emotion-3:hover {
    background: #000000;
  }

  .emotion-3:active {
    background: #000000;
  }
}

<a
  class="emotion-0 emotion-1 emotion-2 emotion-3 emotion-4"
  data-next-link="true"
  data-next-link-legacy-behavior="false"
  href="https://www.makeswift.com/"
  id="test-button"
  target="_self"
>
  Button Text
</a>
`;
